core/src/main/java/com/graphhopper/routing/weighting/custom/ExpressionBuilder.java

```diff
@@ -0,0 +1,382 @@
+                parseBlockStatements());
+        return speedStatements;
+    }
+
+    /**
+     * Parse the expressions from CustomModel relevant for the method getPriority - see createClassTemplate.
+     *
+     * @return the created statements (parsed expressions)
+     */
+    private static List<Java.BlockStatement> createGetPriorityStatements(Set<String> priorityVariables,
+                                                                         CustomModel customModel, EncodedValueLookup lookup) throws Exception {
+        List<Java.BlockStatement> priorityStatements = new ArrayList<>();
+        priorityStatements.addAll(verifyExpressions(new StringBuilder("double value = 1;\n"), "priority_user_statements",
+                priorityVariables, customModel.getPriority(), lookup,
+                num -> "value *= " + num + ";\n", "return value;"));
+        String priorityMethodStartBlock = "";
+        for (String arg : priorityVariables) {
+            priorityMethodStartBlock += getVariableDeclaration(lookup, arg);
+        }
+        priorityStatements.addAll(0, new Parser(new org.codehaus.janino.Scanner("getPriority", new StringReader(priorityMethodStartBlock))).
+                parseBlockStatements());
+        return priorityStatements;
+    }
+
+    static boolean isValidVariableName(String name) {
+        return name.startsWith(AREA_PREFIX) || allowedNames.contains(name);
+    }
+
+    /**
+     * For the methods getSpeed and getPriority we declare variables that contain the encoded value of the current edge
+     * or if an area contains the current edge.
+     */
+    private static String getVariableDeclaration(EncodedValueLookup lookup, String arg) {
+        if (arg.startsWith(AREA_PREFIX)) {
+            String id = arg.substring(AREA_PREFIX.length());
+            return "boolean " + arg + " = " + CustomWeightingHelper.class.getSimpleName() + ".in(this.area_" + id + ", edge);\n";
+        } else if (lookup.hasEncodedValue(arg)) {
+            check(arg);
+            EncodedValue enc = lookup.getEncodedValue(arg, EncodedValue.class);
+            return getReturnType(enc) + " " + arg + " = reverse ? " +
+                    "edge.getReverse((" + getInterface(enc) + ") this." + arg + "_enc) : " +
+                    "edge.get((" + getInterface(enc) + ") this." + arg + "_enc);\n";
+        } else if (isValidVariableName(arg)) {
+            return "";
+        } else {
+            throw new IllegalArgumentException("Not supported " + arg);
+        }
+    }
+
+    /**
+     * @return the interface as string of the provided EncodedValue, e.g. IntEncodedValue (only interface) or BooleanEncodedValue (first interface)
+     */
+    private static String getInterface(EncodedValue enc) {
+        if (enc.getClass().getInterfaces().length == 0) return enc.getClass().getSimpleName();
+        return enc.getClass().getInterfaces()[0].getSimpleName();
+    }
+
+    private static String getReturnType(EncodedValue encodedValue) {
```

Comment:
Could we move this into the EncodedValue interface?

e.g

```java
public interface EncodedValue {
    Class<?> getReturnType();
```

```java
public interface BooleanEncodedValue extends EncodedValue {

    default Class<?> getReturnType() {
        return boolean.class;
    }
```

Reply:
For now I would keep this feature private